### ISSUE TITLE
Docker daemon crash  - 18.03.1-ce - Amazon Linux -  Segmentation-Fault@containerd/dialer/dialer.go:43

### ISSUE THREAD

**Issue by eranb** [docker/for-linux#378]
* [x] This is a bug report
* [ ] This is a feature request
* [x] I searched existing issues before opening this one

### Expected behavior
Docker daemon shouldn't crash when pulling image

### Actual behavior
Docker daemon crash

### Steps to reproduce the behavior
docker pull ECR-PATH

**Output of `docker version`:**
```
Client:
 Version:      18.03.1-ce
 API version:  1.37
 Go version:   go1.9.4
 Git commit:   3dfb8343b139d6342acfd9975d7f1068b5b1c3d3
 Built:        Fri Jul 20 19:24:06 2018
 OS/Arch:      linux/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      18.03.1-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.4
  Git commit:   7390fc6/18.03.1-ce
  Built:        Fri Jul 20 19:25:23 2018
  OS/Arch:      linux/amd64
  Experimental: false
```

**Output of `docker info` (After I restated the service ):**
```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 18.03.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.14.47-56.37.amzn1.x86_64
Operating System: Amazon Linux AMI 2018.03
OSType: linux
Architecture: x86_64
CPUs: 16
Total Memory: 30.42GiB
Name: ip-10-10-30-116
ID: OQHT:ZKSK:7QU4:I3QF:455S:BV6Y:VTLK:J4G2:73CB:CDXU:LBTT:BKFF
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

```

**Additional environment details (AWS, VirtualBox, physical, etc.)**
```
[root@XXX~]# uname -a
Linux XXX 4.14.47-56.37.amzn1.x86_64 #1 SMP Wed Jun 6 18:49:01 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
[root@XXX ~]# yum list installed |grep docker
docker.x86_64                        18.03.1ce-5.0.amzn1           @amzn-updates
```
**Docker log**
```
time="2018-07-28T09:15:36Z" level=info msg="shim reaped" id=cb9206a4894e2997c240070dd1f886ce2f942c66f62928ab624a06690c21de9d module="containerd/tasks"
time="2018-07-28T09:15:36.942836940Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
time="2018-07-29T08:21:10Z" level=info msg="shim docker-containerd-shim started" address="/containerd-shim/moby/141b3e1b204d9d95072a40e91ce627a279aca244c970132e51c53bbd8ccacba8/shim.sock" debug=false module="containerd/tasks" pid=13890
time="2018-07-29T08:33:11Z" level=info msg="shim reaped" id=141b3e1b204d9d95072a40e91ce627a279aca244c970132e51c53bbd8ccacba8 module="containerd/tasks"
time="2018-07-29T08:33:11.087779485Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
time="2018-07-29T18:00:13Z" level=info msg="shim docker-containerd-shim started" address="/containerd-shim/moby/d7335a385613176e0f2baa5af12c0c76a257c74626c04c779fea667dc728227c/shim.sock" debug=false module="containerd/tasks" pid=8734
time="2018-07-29T18:01:38Z" level=info msg="shim reaped" id=d7335a385613176e0f2baa5af12c0c76a257c74626c04c779fea667dc728227c module="containerd/tasks"
time="2018-07-29T18:01:38.380241174Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
time="2018-07-29T18:02:46Z" level=info msg="shim docker-containerd-shim started" address="/containerd-shim/moby/36c8b205156aa7cb6b4425bc170320990314a8488fa2b351b36228d7c613ed81/shim.sock" debug=false module="containerd/tasks" pid=19331
time="2018-07-29T18:03:28Z" level=info msg="shim reaped" id=36c8b205156aa7cb6b4425bc170320990314a8488fa2b351b36228d7c613ed81 module="containerd/tasks"
time="2018-07-29T18:03:28.013904407Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
time="2018-07-29T21:47:43Z" level=info msg="shim docker-containerd-shim started" address="/containerd-shim/moby/a7f43bb0a2e902a5d12158595cfb832ba0a28469d27220750f1340807e6a7e1a/shim.sock" debug=false module="containerd/tasks" pid=29088
time="2018-07-29T22:00:34Z" level=info msg="shim reaped" id=a7f43bb0a2e902a5d12158595cfb832ba0a28469d27220750f1340807e6a7e1a module="containerd/tasks"
time="2018-07-29T22:00:34.597151952Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x55f5df515c4b]

goroutine 1830 [running]:
github.com/docker/docker/vendor/github.com/containerd/containerd/dialer.Dialer.func2(0xc420b58000)
        /builddir/build/BUILD/docker-ce-18.03.1-ce/components/engine/_build/src/github.com/docker/docker/vendor/github.com/containerd/containerd/dialer/dialer.go:46 +0x5b
created by github.com/docker/docker/vendor/github.com/containerd/containerd/dialer.Dialer
        /builddir/build/BUILD/docker-ce-18.03.1-ce/components/engine/_build/src/github.com/docker/docker/vendor/github.com/containerd/containerd/dialer/dialer.go:43 +0x1c9
```

**Comment by thaJeztah** [comment]
Looks like this issue https://github.com/containerd/containerd/pull/2022 / https://github.com/containerd/containerd/pull/2034 (which also has the fix)

Looking at 18.03.1 CE, I see that version of the containerd client does not yet have that fix; https://github.com/docker/docker-ce/blob/v18.03.1-ce/components/engine/vendor/github.com/containerd/containerd/dialer/dialer.go#L43-L48

(there's an open PR for that; https://github.com/docker/docker-ce/pull/523, but not sure if a 18.03.2 patch release will be done)

**Comment by eranb** [comment]
Thanks @thaJeztah,
do you know stable version of Docker (w/o the bug) that I could downgrade to ?

**Comment by thaJeztah** [comment]
18.06 is current stable, but there's currently no official packages for Amazon Linux (those are built and supported by Amazon)

**Comment by eranb** [comment]
Amazon released new stable of CE `18.06.1` ( with the fix ) for Amazon Linux
